src/main/java/net/minecraftforge/common/TierSortingRegistry.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.graph.ElementOrder;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.tags.Tag;
+import net.minecraft.world.item.Tier;
+import net.minecraft.world.item.Tiers;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraftforge.fml.loading.toposort.TopologicalSort;
+
+import javax.annotation.Nullable;
+import java.util.*;
+
+public class TierSortingRegistry
+{
+    public static synchronized void registerTier(Tier tier, ResourceLocation name, List<Object> after, List<Object> before)
+    {
+        if (tiers.containsKey(name))
+            throw new IllegalStateException("Duplicate tier name " + name);
+
+        processTier(tier, name, after, before);
+    }
+
+    public static List<Tier> getSortedTiers()
```

Comment:
Can this list be cached by modders? If so, when is it a safe time to do so?

Reply:
Ah good reminder. I meant to change it to be a final list that gets refilled as needed.

Reply:
As of the latest push, the list updates happen always in either the server thread (loading/reloading), or the client thread (sync received), so it should be safe to keep a copy of the list, so long as you don't keep an iterator. :P